### PR TITLE
PR for vision branch - working note detection and new drive method. 

### PR DESCRIPTION
**Pull request by SJJCoding**

### CODE REVIEW

File: src/main/java/frc/team2412/robot/Controls.java [hunk 1]
```diff
@@ -66,6 +69,13 @@ public Controls(Subsystems s) {
 		if (DRIVEBASE_ENABLED) {
 			bindDrivebaseControls();
 		}
+		if (LIMELIGHT_ENABLED) {
+			bindLimelightControls();
+		}
+		// if (LAUNCHER_ENABLED) {
+		//	bindLauncherControls();
+		// }
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/Controls.java]: Delete commented out code

File: src/main/java/frc/team2412/robot/Controls.java [hunk 2]
```diff
@@ -95,10 +105,12 @@ private void bindDrivebaseControls() {
 								driveController::getLeftY,
 								driveController::getLeftX,
 								() -> Rotation2d.fromRotations(driveController.getRightX())));
-		driveController.start().onTrue(new InstantCommand(s.drivebaseSubsystem::resetGyro));
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/Controls.java:119]: Bad merge? resetGyro needs to be mapped to something

File: src/main/java/frc/team2412/robot/Subsystems.java [hunk 3]
```diff
@@ -14,7 +15,7 @@ public static class SubsystemConstants {
 		private static final boolean IS_COMP = Robot.getInstance().isCompetition();
 
 		public static final boolean APRILTAGS_ENABLED = false;
-		public static final boolean LIMELIGHT_ENABLED = false;
+		public static final boolean LIMELIGHT_ENABLED = true;
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/Subsystems.java]: Should stay `false`

File: src/main/java/frc/team2412/robot/commands/drivebase/DriveCommand.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+			DrivebaseSubsystem drivebaseSubsystem,
+			DoubleSupplier forward,
+			DoubleSupplier strafe,
+			DoubleSupplier rotation,
+			DoubleSupplier turboRotation) {
+		this.drivebaseSubsystem = drivebaseSubsystem;
+		this.forward = forward;
+		this.strafe = strafe;
+		this.rotation = rotation;
+		// this variable give the right trigger input
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/commands/drivebase/DriveCommand.java]: Delete, as this will quickly get out of sync with the control mapping

File: src/main/java/frc/team2412/robot/commands/drivebase/DriveCommand.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+		// math for normalizing and cubing inputs
+		double magnitude = Math.pow(Math.min(Math.sqrt(Math.pow(x, 2) + Math.pow(y, 2)), 1), 3);
+		double angle = Math.atan2(y, x);
+		double cubed_x = magnitude * Math.cos(angle);
+		double cubed_y = magnitude * Math.sin(angle);
+
+		drivebaseSubsystem.simpleDrive(
+				cubed_x * driveSpeedEntry.getDouble(1.0) * 4.4196, // convert from percent to m/s
+				cubed_y * driveSpeedEntry.getDouble(1.0) * 4.4196,
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/commands/drivebase/DriveCommand.java]: The drivebases all have different max velocities, so this should be robot-dependent.

File: src/main/java/frc/team2412/robot/commands/limelight/GetWithinDistanceCommand.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
+/*package frc.team2412.robot.commands.limelight;
+
+
+public class GetWithinDistanceCommand extends Command {
+
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/commands/limelight/GetWithinDistanceCommand.java]: Consider uncommenting this, and just not using the command yet.

File: src/main/java/frc/team2412/robot/commands/limelight/GetWithinDistanceCommand.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+	private Command getToPositionCommand;
+
+	public GetWithinDistanceCommand(
+			LimelightSubsystem limelightSubsystem, DrivebaseSubsystem drivebaseSubsystem) {
+		this.limelightSubsystem = limelightSubsystem;
+		this.drivebaseSubsystem = drivebaseSubsystem;
+
+		// find target -> get current pose -> use current pose and target to calculate target pose ->
+		// move towards target pose
+
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/commands/limelight/GetWithinDistanceCommand.java]: Need to require the drivebaseSubsystem so nothing else tries to drive while this command is running

File: src/main/java/frc/team2412/robot/commands/limelight/GetWithinDistanceCommand.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+		getToPositionCommand.execute();
+		limelightSubsystem.getTargetPose(currentPose);
+	}
+
+	@Override
+	public void end(boolean interrupted) {}
+
+	@Override
+	public boolean isFinished() {
+		System.out.println("within distance");
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/commands/limelight/GetWithinDistanceCommand.java]: Delete

File: src/main/java/frc/team2412/robot/subsystems/DrivebaseSubsystem.java [hunk 9]
```diff
@@ -164,6 +164,20 @@ public void drive(Translation2d translation, Rotation2d rotation, boolean fieldO
 		}
 	}
 
+	public void driveWithDoubleSupplierTurn(
+			Translation2d translation, Supplier<Double> rotation, boolean fieldOriented) {
+		// if we're requesting the robot to stay still, lock wheels in X formation
+		if (translation.getNorm() == 0 && rotation.get() == 0.0 && xWheelsEnabled) {
+			swerveDrive.lockPose();
+		}
+		if (rotationSetpoint != null) {
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystems/DrivebaseSubsystem.java]: Make this `else if` as the robot shouldn't try to move while the wheel are in the 'lockPose' position.

File: src/main/java/frc/team2412/robot/subsystems/DrivebaseSubsystem.java [hunk 10]
```diff
@@ -193,6 +207,14 @@ public Command driveJoystick(
 				});
 	}
 
+	public void simpleDrive(double forward, double strafe, Rotation2d rotation) {
+
+		ChassisSpeeds chassisSpeeds = new ChassisSpeeds(0, 0, 0);
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystems/DrivebaseSubsystem.java]: Delete unnecessary initialization to `ChassisSpeeds`.

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+	public static final double TARGET_HEIGHT = 0.33;
+
+	public static final double GOAL_DISTANCE_FROM_TARGET = 0.7;
+	public static final double GOAL_DISTANCE_FROM_CONE = 0.3;
+	public static final double GOAL_DISTANCE_FROM_CUBE = 0.3;
+
+	// MEMBERS
+
+	NetworkTable networkTable;
+
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: Add `final` since this never changes

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+	String currentPoseString;
+	String targetPoseString;
+
+	// network tables
+
+	// CONSTRUCTOR !
+	public LimelightSubsystem() {
+
+		// broadcast 10.24.12.227
+
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: ? Not sure what this is for. The Limelight should have a static IP address, likely 10.24.12.11 (Orange Pi is at 10.24.12.12)

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+				.withSize(4, 1);
+
+		limelightTab
+				.addString("Target Pose ", this::getTargetPoseString)
+				.withPosition(0, 2)
+				.withSize(4, 1);
+	}
+
+	public void setPipeline() {
+		networkTable.getEntry("pipeline").setNumber(2);
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: Can be `private` as nothing outside the class calls this.

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+		return networkTable.getEntry("tshort").getDouble(0);
+	}
+
+	public double getDistanceFromTarget() {
+
+		// 1.3 at 1.9
+		// 1.0 at 1.3
+
+		double angleToTarget = getVerticalOffset();
+		// return (TARGET_HEIGHT - CAMERA_MOUNT_HEIGHT) / Math.tan(CAMERA_ANGLE_OFFSET + angleToTarget);
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: Delete commented-out code, or add a comment describing why it's still here

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+		} else {
+			return 0.0;
+		}
+	}
+
+	// tan(degree) * distance = sideways distance
+
+	// target height / tan(vertical angle)
+
+	// TODO fix this or something
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: Please add more about why this is broken, and what needs fixing

**Review by jamesdooley4 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/team2412/robot/Controls.java (hunk 2 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/Controls.java:119]: Not sure what it should be mapped to since the start is taken up by testing, can make a boolean or something that toggles it

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java (hunk 15 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: done

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java (hunk 14 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: done

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java (hunk 13 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: done

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java (hunk 12 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: Changed the comment to describe the location of the limelight config and added proper address

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java (hunk 11 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: done

File: src/main/java/frc/team2412/robot/subsystems/DrivebaseSubsystem.java (hunk 10 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/subsystems/DrivebaseSubsystem.java]: done

File: src/main/java/frc/team2412/robot/subsystems/DrivebaseSubsystem.java (hunk 9 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/subsystems/DrivebaseSubsystem.java]: done

File: src/main/java/frc/team2412/robot/commands/limelight/GetWithinDistanceCommand.java (hunk 8 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/commands/limelight/GetWithinDistanceCommand.java]: The whole GetWithinDistanceCommand.java file is redundant and has been removed

File: src/main/java/frc/team2412/robot/commands/limelight/GetWithinDistanceCommand.java (hunk 7 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/commands/limelight/GetWithinDistanceCommand.java]: The whole GetWithinDistanceCommand.java file is redundant and has been removed

File: src/main/java/frc/team2412/robot/commands/limelight/GetWithinDistanceCommand.java (hunk 6 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/commands/limelight/GetWithinDistanceCommand.java]: The whole GetWithinDistanceCommand.java file is redundant and has been removed

File: src/main/java/frc/team2412/robot/Controls.java (hunk 1 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/Controls.java]: done

File: src/main/java/frc/team2412/robot/Controls.java (hunk 2 above)
Comment (KangarooKoala) [reply src/main/java/frc/team2412/robot/Controls.java:119]: Probably should just talk with drive team about what button would work best for `getWithinDistanceTrigger`, or if that's just for testing then take an unused button.

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java (hunk 15 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: (removed the TODO as it is fixed)

File: src/main/java/frc/team2412/robot/Controls.java (hunk 2 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/Controls.java:119]: okay, will do that in the next set of changes since its just a testing thing

File: src/main/java/frc/team2412/robot/commands/drivebase/DriveCommand.java
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/commands/drivebase/DriveCommand.java:1]: As I recall, this was the command that processed driver input last year, so it has some features (like cubing, deadband, turbo rotation, and sliders) that don't really belong to an automatic command. (Deadband and input cubing especially could mess with making the code control the robot how you want). Should these be removed? (See also the comment about using `drivebaseSubsystem.rotateToAngle()`- Is this command even necessary?)

File: src/main/java/frc/team2412/robot/subsystems/DrivebaseSubsystem.java [hunk 16]
```diff
@@ -164,6 +164,20 @@ public void drive(Translation2d translation, Rotation2d rotation, boolean fieldO
 		}
 	}
 
+	public void driveWithDoubleSupplierTurn(
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/subsystems/DrivebaseSubsystem.java]: What's the point in this method? Nothing seems to call it, and it behaves identically to the `drive()` method above it (except the bug fix, which should be handled in a separate PR).

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+public class LimelightSubsystem extends SubsystemBase {
+
+	// CONSTANTS
+
+	private static final PIDController TRANSLATION_PID = new PIDController(10.0, 0, 0);
+	private static final PIDController ROTATION_PID = new PIDController(8.0, 0, 0);
+
+	// meters?
+	public static final double CAMERA_MOUNT_HEIGHT = 0.1143;
+	public static final double CAMERA_ANGLE_OFFSET = 0;
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: Is this accurate? Looking at the physical setup on this year's robot, the camera is at an angle.

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+		Rotation2d targetHeading =
+				new Rotation2d(
+						currentPose.getRotation().getRadians() + Units.degreesToRadians(getHorizontalOffset()));
+		double targetDistance = getDistanceFromTarget() / 39.3700787;
+
+		double targetX = Math.sin(targetHeading.getRadians()) * targetDistance;
+		double targetY = Math.cos(targetHeading.getRadians()) * targetDistance;
+
+		Pose2d targetPose =
+				new Pose2d(currentPose.getX() + targetY, currentPose.getY() + targetX, targetHeading);
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]:
```suggestion
        Translation2d translationOffset = new Translation2d(targetDistance, targetHeading);

		Pose2d targetPose =
				new Pose2d(currentPose.getTranslation().plus(translationOffset), targetHeading);
```
The variables had the wrong name earlier (`targetX` and `targetY` were swapped), and this also more concisely expresses the logic.

File: src/main/java/frc/team2412/robot/subsystems/DrivebaseSubsystem.java [hunk 19]
```diff
@@ -193,6 +207,13 @@ public Command driveJoystick(
 				});
 	}
 
+	public void simpleDrive(double forward, double strafe, Rotation2d rotation) {
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/subsystems/DrivebaseSubsystem.java]: Where is this method used?

File: src/main/java/frc/team2412/robot/Subsystems.java [hunk 20]
```diff
@@ -24,6 +25,7 @@ public static class SubsystemConstants {
 	public final DrivebaseWrapper drivebaseWrapper;
 	public final DrivebaseSubsystem drivebaseSubsystem;
 	public final LauncherSubsystem launcherSubsystem;
+	public LimelightSubsystem limelightSubsystem;
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/Subsystems.java]:
```suggestion
	public final LimelightSubsystem limelightSubsystem;
```
(For consistency with the other subsystems)

File: src/main/java/frc/team2412/robot/Subsystems.java [hunk 21]
```diff
@@ -46,6 +48,9 @@ public Subsystems() {
 		} else {
 			launcherSubsystem = null;
 		}
+		if (LIMELIGHT_ENABLED) {
+			limelightSubsystem = new LimelightSubsystem();
+		}
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/Subsystems.java:56]:
```suggestion
		} else {
		    limelightSubsystem = null;
		}
```
(For consistency with the other subsystems)

File: src/main/java/frc/team2412/robot/Controls.java [hunk 22]
```diff
@@ -95,10 +101,12 @@ private void bindDrivebaseControls() {
...
 		driveController.rightStick().onTrue(new InstantCommand(s.drivebaseSubsystem::toggleXWheels));
 	}
 
+	public void bindLimelightControls() {
+		getWithinDistanceTrigger.onTrue(s.limelightSubsystem.getWithinDistance(s.drivebaseSubsystem));
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/Controls.java]:
```suggestion
		getWithinDistanceTrigger.onTrue(
		        Commands.deferred(
		                () -> s.limelightSubsystem.getWithinDistance(s.drivebaseSubsystem),
		                Set.of(s.limelightSubsystem, s.drivebaseSubsystem)));
```
Otherwise, `getWithinDistance()` would only be called once, so `hasTargets()` would only be checked once (at the start of robot code). Alternately, make the supplier used by `getWithinDistance()` check `hasTargets()` (a ternary operator would be useful here), so that `hasTargets()` will be repeatedly checked.

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java [hunk 23]
```diff
@@ -0,0 +1,196 @@
...
+	}
+
+	public Command getWithinDistance(DrivebaseSubsystem drivebaseSubsystem) {
+		final DoubleSupplier returnTurn = () -> turnPowerLin();
+		final DoubleSupplier returnZero = () -> 0.0;
+		Command moveCommand;
+		if (hasTargets()) {
+			System.out.println("has targets");
+			moveCommand =
+					new DriveCommand(drivebaseSubsystem, returnZero, returnZero, returnTurn, returnZero);
```
Comment (jbko6) [inline src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: Why are you using DriveCommand instead of DrivebaseSubsystem.rotateToAngle?

Comment (jbko6) [reply src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: If the issue with rotateToAngle is that it expects a field relative rotation instead of robot relative rotation, you could add add turnPowerLin() to the current robot rotation and emulate the behavior of DriveCommand in your code.

Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: mostly because the end goal is to have it rotate and drive at the same time for better efficiency. Driving is just turned off not to cause osha violations

File: src/main/java/frc/team2412/robot/subsystems/DrivebaseSubsystem.java (hunk 16 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/subsystems/DrivebaseSubsystem.java]: Probably tested it and it didn't work, will remove

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java (hunk 23 above)
Comment (jbko6) [reply src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: rotate and drive as in drive to the note? or rotate and drive as in rotate as the driver is driving? If you're trying to go for the first option, PathPlanner has functions that can generate paths like that for you. If you're going for the second option, rotateToAngle is built to handle that. Either way, adding a whole new DriveCommand (which will get confused with DrivebaseSubsystem's driveJoystick) seems like a very complex way to solve this issue

File: src/main/java/frc/team2412/robot/subsystems/DrivebaseSubsystem.java (hunk 19 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/subsystems/DrivebaseSubsystem.java]: In drivecommand. This method will be deleted once drivecommand is switched to a better drive method

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java (hunk 23 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: Rotate and drive to the note simultaneously. Kirby and I tried to get pathplanner to generate paths in that way and it never worked. Either way, drivecommand will probably be changed in favor of just using either drive() or drivejoystick().

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java (hunk 18 above)
Comment (KangarooKoala) [reply src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: (The suggested change was undone in ae0af- I'm guessing the device that pushed that commit hadn't pulled in the commit from the suggestion)

File: src/main/java/frc/team2412/robot/Controls.java (hunk 22 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/Controls.java]: Outdated because of new push and logic restructuring

File: src/main/java/frc/team2412/robot/Subsystems.java (hunk 21 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/Subsystems.java:56]: Done

File: src/main/java/frc/team2412/robot/Subsystems.java (hunk 20 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/Subsystems.java]: Done

File: src/main/java/frc/team2412/robot/commands/drivebase/DriveCommand.java
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/commands/drivebase/DriveCommand.java:1]: removed the old method and replaced it with DriveToNoteCommand(), which uses the drivebase drive() method

File: src/main/java/frc/team2412/robot/commands/drivebase/DriveCommand.java (hunk 4 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/commands/drivebase/DriveCommand.java]: Outdated because of new push and logic restructuring

File: src/main/java/frc/team2412/robot/commands/drivebase/DriveCommand.java (hunk 5 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/commands/drivebase/DriveCommand.java]: Outdated because of new push and logic restructuring

File: src/main/java/frc/team2412/robot/Subsystems.java (hunk 3 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/Subsystems.java]: Done

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java (hunk 17 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: Distance finder no longer requires these values, removed

File: src/main/java/frc/team2412/robot/subsystems/DrivebaseSubsystem.java (hunk 16 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/subsystems/DrivebaseSubsystem.java]: Removed

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java (hunk 23 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: The new DriveToNoteCommand() works and uses the drive() method

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java (hunk 18 above)
Comment (SJJCoding) [reply src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: re-added

File: src/main/java/frc/team2412/robot/Controls.java (hunk 2 above)
Comment (TAKBS2412) [reply src/main/java/frc/team2412/robot/Controls.java:119]: Can we add this back and move the get within distance trigger to somewhere else?

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+	// target height / tan(vertical angle)
+
+	public Pose2d getTargetPose(Pose2d currentPose) {
+
+		// math thing to get target pose using current pose
+
+		Rotation2d targetHeading =
+				new Rotation2d(
+						currentPose.getRotation().getRadians() + Units.degreesToRadians(getHorizontalOffset()));
+		double targetDistance = getDistanceFromTarget() / 39.3700787;
```
Comment (TAKBS2412) [inline src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: FYI that wpilib has a built in method for converting inches to meters: https://github.wpilib.org/allwpilib/docs/release/java/edu/wpi/first/math/util/Units.html#inchesToMeters(double)

File: src/main/java/frc/team2412/robot/commands/drivebase/DriveToNoteCommand.java [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+		DrivebaseSubsystem drivebaseSubsystem, LimelightSubsystem limelightSubsystem) {
+		this.drivebaseSubsystem = drivebaseSubsystem;
+		this.limelightSubsystem = limelightSubsystem;
+		addRequirements(drivebaseSubsystem);
+	}
+
+	@Override
+	public void execute() {
+		Translation2d move =
+				new Translation2d(INVERT_DRIVE_DIRECTION * limelightSubsystem.getDistanceFromTarget() / INCHES_TO_METERS, 0.0);
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/commands/drivebase/DriveToNoteCommand.java]: Use https://github.wpilib.org/allwpilib/docs/release/java/edu/wpi/first/math/util/Units.html#inchesToMeters(double) here as well

File: src/main/java/frc/team2412/robot/commands/drivebase/DriveToNoteCommand.java [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+		Rotation2d turn = new Rotation2d().fromDegrees(2 * INVERT_DRIVE_DIRECTION * limelightSubsystem.getHorizontalOffset());
+		drivebaseSubsystem.drive(move, turn, false);
+	}
+
+	@Override
+	public boolean isFinished() {
+		return (limelightSubsystem.getDistanceFromTarget() <= 20);
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/commands/drivebase/DriveToNoteCommand.java]: The `20` here should at least be a named constant (with units in the name). Ideally it's a shuffleboard-configurable value so it can be quickly tweaked during testing without having to redeploy robot code.

Comment (jamesdooley4) [reply src/main/java/frc/team2412/robot/commands/drivebase/DriveToNoteCommand.java]: Actually, this should just be `limelightSubsystem.isWithinDistance()` so the same constant isn't in two different files.

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+import edu.wpi.first.networktables.NetworkTable;
+import edu.wpi.first.networktables.NetworkTableInstance;
+import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard;
+import edu.wpi.first.wpilibj.shuffleboard.ShuffleboardTab;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+public class LimelightSubsystem extends SubsystemBase {
+
+	// POTENTIAL CONSTANTS
+	public static final double GOAL_DISTANCE_FROM_NOTE = 20.0;
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: Ideally this constant would just be the default value, and the actual value used in `isWithinDistance()` is a Shuffleboard-configurable value, so it's fast and easy to change it and retest without having to redeploy robot code.

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java [hunk 28]
```diff
@@ -0,0 +1,133 @@
...
+	String currentPoseString;
+	String targetPoseString;
+
+	// network tables
+
+	// CONSTRUCTOR !
+	public LimelightSubsystem() {
+
+		// camera stream at http://10.24.12.21:5800
+
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: The Limelight should use an available static IP address between .10 and .20 (ideally 10.24.12.11). .12 is taken by the Orange PI

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java [hunk 29]
```diff
@@ -0,0 +1,133 @@
...
+	}
+	// TODO re-measure distances to make this right
+	public double getDistanceFromTarget() {
+
+		// focal length = (P x D) / W
+		double focal_length = 559.0394;
+
+		// distance = (W x F) / P
+		// returns inches for testing purposes, will divide by 39.3700787 to return meters
+		return (14 * focal_length) / getBoxWidth();
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: To avoid unit confusion, please update the method name to include the units it returns (Inches), so it's obvious that callers need to be updated if this switches to meters.

**Review by jamesdooley4 — CHANGES_REQUESTED** [review]

**Comment by jamesdooley4** [comment]
Overall, this is looking a lot better & I'm hoping to be able to see it in use where the robot just automatically finds & grabs the note in front of it.

File: src/main/java/frc/team2412/robot/Controls.java [hunk 30]
```diff
@@ -118,6 +129,14 @@ private void bindDrivebaseControls() {
 		// 										new Pose2d(new Translation2d(1.3, 5.55), new Rotation2d(180)))));
 	}
 
+	// leaving this code in in case we need to test outside of auto
+	/*
+	public void bindLimelightControls() {
+		getWithinDistanceTrigger.onTrue(
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/Controls.java]: Using `whileTrue` would be safer since the driver can cancel the command just by releasing the button.

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java [hunk 31]
```diff
@@ -0,0 +1,146 @@
...
+	}
+	// TODO re-measure distances to make this right
+	public double getDistanceFromTargetInches() {
+
+		// focal length = (P x D) / W
+		double focal_length = 559.0394;
+
+		// distance = (W x F) / P
+		// returns inches for testing purposes, will divide by 39.3700787 to return meters
+		return (14 * focal_length) / getBoxWidth();
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java]: This divides by zero if there's no target or the limelight isn't publishing (default of box width is 0). That might be fine if all callers always verify `hasTargets()` returns true. Or consider updating this to return `Optional<Double>` to make it clear that sometimes it'll return "don't know".

File: src/main/java/frc/team2412/robot/commands/drivebase/DriveToNoteCommand.java [hunk 32]
```diff
@@ -0,0 +1,42 @@
...
+	public void execute() {
+		Translation2d move =
+				new Translation2d(
+						INVERT_DRIVE_DIRECTION
+								* Units.inchesToMeters(limelightSubsystem.getDistanceFromTargetInches()),
+						0.0);
+		Rotation2d turn =
+				new Rotation2d()
+						.fromDegrees(2 * INVERT_DRIVE_DIRECTION * limelightSubsystem.getHorizontalOffset());
+		drivebaseSubsystem.drive(move, turn, false);
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/commands/drivebase/DriveToNoteCommand.java]: Should likely check that there's a target before driving, so it doesn't keep running when there's nothing visible

File: src/main/java/frc/team2412/robot/commands/drivebase/DriveToNoteCommand.java [hunk 33]
```diff
@@ -0,0 +1,42 @@
...
+	}
+
+	@Override
+	public boolean isFinished() {
+		return (limelightSubsystem.isWithinDistance());
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/commands/drivebase/DriveToNoteCommand.java]: Should this also complete if there's no longer a target (or only end if there is a target & it's within distance)

File: src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java [hunk 34]
```diff
@@ -0,0 +1,146 @@
...
+	}
+
+	public String getTargetPoseString() {
+		return targetPoseString;
+	}
+
+	public boolean isWithinDistance() {
+		return (getDistanceFromTargetInches() <= GOAL_DISTANCE_FROM_NOTE.getDouble(20.0));
+	}
+
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java:109]: Should this check if there's a visible target first?

**Review by jamesdooley4 — CHANGES_REQUESTED** [review]
Needs to handle case there's no target visible. Other than that, it looks good.

**Comment by SJJCoding** [comment]
ready to merge fr fr

File: src/main/java/frc/team2412/robot/commands/drivebase/DriveToNoteCommand.java [hunk 35]
```diff
@@ -24,19 +24,21 @@ public DriveToNoteCommand(
...
+					new Rotation2d()
+							.fromDegrees(2 * INVERT_DRIVE_DIRECTION * limelightSubsystem.getHorizontalOffset());
+			drivebaseSubsystem.drive(move, turn, false);
+		}
 	}
 
 	@Override
 	public boolean isFinished() {
-		return (limelightSubsystem.isWithinDistance());
+		return (limelightSubsystem.isWithinDistance() || !limelightSubsystem.hasTargets());
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/commands/drivebase/DriveToNoteCommand.java:42]: Should be `&&` ?  Else it'll end as soon as a note is seen.

**Review by jamesdooley4 — APPROVED** [review]
